pkg/agent/controller/networkpolicy/priority.go

```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+		} else {
+			emptiedSlotsHigh++
+			// visit costMap in the reverse direction
+			mapIndex := target - emptiedSlotsHigh
+			c, ok := costMap[mapIndex]
+			// only add to the costMap if the counterpart cost is available. i.e. if the target is 4, and cost for
+			// using 2 empty slots high is computed, it does not make sense to store this cost if there's no entry
+			// for cost that uses 2 empty slots low (indicating no 2 empty slots can be found starting from lowerBound).
+			if ok {
+				c.cost = costSiftDown + costSiftUp
+				c.upperBound = reassignBoundHigh
+			} else if mapIndex == 0 {
+				costMap[mapIndex] = &reassignCost{lowerBound + 1, reassignBoundHigh, costSiftUp}
+			}
 		}
+		reassignBoundHigh++
 	}
-	return nil
-}
-
-// getLastOccupiedOFPriority returns the first ofPriority lower than the input ofPriority that is
-// currently occupied in the table, as well as the corresponded Priority.
-// Note that the function must return a ofPriority that is lower than the input ofPriority.
-// The search is based on sortedOFPriorities as it assumes the table is sparse in most cases.
-func (pa *priorityAssigner) getLastOccupiedOFPriority(ofPriority uint16) (*uint16, *types.Priority) {
-	idx := sort.Search(len(pa.sortedOFPriorities), func(i int) bool { return ofPriority <= pa.sortedOFPriorities[i] })
-	if idx > 0 {
-		lastOccupied := pa.sortedOFPriorities[idx-1]
-		priority := pa.ofPriorityMap[lastOccupied]
-		return &lastOccupied, &priority
+	minCost, minCostIndex := math.MaxInt32, 0
+	for i := target; i >= 0; i-- {
+		if cost, exists := costMap[i]; exists && cost.cost < minCost {
+			// make sure that the reassign range makes up for all Priorities to be registered.
+			if int(cost.upperBound-cost.lowerBound)+1 == numNewPriorities+cost.cost {
+				minCost = cost.cost
+				minCostIndex = i
+			}
+		}
 	}
-	return nil, nil
+	if minCost == math.MaxInt32 {
+		// theoretically this should not happen since Priority overflow is checked earlier.
+		return lowerBound, upperBound, fmt.Errorf("failed to push boundary priorities to reach numNewPriorities")
+	}
+	return costMap[minCostIndex].lowerBound, costMap[minCostIndex].upperBound, nil
 }
 
-// getLastVacantOFPriority returns the first lower ofPriority that is currently vacant in the table,
-// starting from the ofPriority one below the input.
-// The search is incrementally against all ofPriorities available as it assumes the table is sparse in most cases.
-func (pa *priorityAssigner) getLastVacantOFPriority(ofPriority uint16) *uint16 {
-	for i := ofPriority - 1; i >= PolicyBottomPriority; i-- {
-		// ofPriority-1 will be less than or equal to PolicyTopPriority
-		if _, exists := pa.ofPriorityMap[i]; !exists {
-			return &i
+// reassignBoundaryPriorities reassigns Priorities from lowerBound / upperBound or both, to make room for
+// new Priorities to be registered. It also records all the priority updates due to the reassignment in the
+// map updates which is passed to it as parameter.
```

Comment:
s/map updates/updates map

or

s/map updates/map of updates

Reply:
Done.